sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -264,6 +264,9 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       val options = sessionOptions ++ extraOptions + checkFilesExistsOption
       val dsOptions = new DataSourceOptions(options.asJava)
       provider.getTable(dsOptions) match {
+        case table: SupportsDirectWrite =>
```

Comment:
this should work without save mode. That said, we should add a new flag in `AppendData` and other operators to indicate if it needs schema validation.

Reply:
Creating `DataSourceV2Relation` requires table schema. Take file source as an example, Spark doesn't need to infer data schema here.